### PR TITLE
Add `Http1HeaderNaming` rule for WebClient

### LINKED ISSUE #3196: Provide an option to Armeria client to keep the original header key
We need a possibility to Armeria client not to convert headers to lowercase for HTTP/1, because some servers violate case insensitive HTTP/1 requirements.

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

Armeria normalizes HTTP headers with lower cases for performance.
However, insensitive HTTP headers are rejected by some HTTP/1 servers.
See #3196 for more information.

Modifications:

- Add `Http1HeaderNaming` rule for customizing HTTP/1 header names.
- Add `http1HeaderNaming` to `ClientFactoryBuilder`
- Use the injected `http1HeaderNaming` in `ClientHttp1ObjectEncoder`

Result:

- You can now set your own HTTP/1 header naming rule using `ClientFactoryBuilder.http1HeaderNaming()`.
- Fixes #3196

### CODE REVIEW

**Review by minwoox — APPROVED** [review]
Thanks!

File: core/src/main/java/com/linecorp/armeria/common/HttpHeaderNames.java [hunk 1]
```diff
@@ -70,6 +70,7 @@
     private static final BitSet PROHIBITED_NAME_CHARS;
     private static final String[] PROHIBITED_NAME_CHAR_NAMES;
     private static final byte LAST_PROHIBITED_NAME_CHAR;
+    private static final ImmutableMap.Builder<AsciiString, String> inverseMapBuilder = ImmutableMap.builder();
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/HttpHeaderNames.java:73]: Could we set `null` to this field after calling `.build()`, so we're sure nobody calls it after that point?

File: core/src/main/java/com/linecorp/armeria/common/Http1HeaderNaming.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Converts a normalized HTTP/2 header name to another HTTP/1 header name.
+ */
+@FunctionalInterface
+public interface Http1HeaderNaming {
+
+    /**
+     * Converts lower-cased HTTP/2 header names to the traditional HTTP/1 header names which are defined at
+     * {@link HttpHeaderNames}. For example, {@code "user-agent"} is converted to {@code "User-Agent"}.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/Http1HeaderNaming.java:31]: Could you make it clear that the headers not defined there will be sent in lower cases?

**Review by trustin — APPROVED** [review]
Thanks!

**Comment by trustin** [comment]
@ikhoon Are you gonna work on the server side as well, so that bad clients are worked around? :laughing:

**Pull request merged**